File: backend/app/repositories/emergency_repository.py
"""Emergency repository — in-memory (demo) and Supabase (live) implementations."""

from __future__ import annotations

import logging
from typing import Any, Protocol

from app.core.config import Settings, get_settings
from app.core.exceptions import ApiError, ErrorCode
from app.db.supabase import get_supabase_admin_client
from app.repositories.state import state
from app.schemas.emergency import (
    EmergencyAllocationData,
    EmergencyAllocationResponse,
    EmergencyRequestData,
    EmergencyRequestResponse,
    EmergencySummary,
)
from app.schemas.common import now_utc

logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# Protocol
# ---------------------------------------------------------------------------


class EmergencyRepository(Protocol):
    def create_request(self, consumer_id: str, consumer_name: str, data: Any) -> EmergencyRequestResponse: ...
    def get_my_requests(self, consumer_id: str) -> list[EmergencyRequestResponse]: ...
    def get_all_requests(self) -> list[EmergencyRequestResponse]: ...
    def get_request(self, request_id: str) -> EmergencyRequestResponse | None: ...
    def update_request(self, request_id: str, update: Any, admin_id: str) -> EmergencyRequestResponse: ...
    def create_allocation(self, data: Any, allocated_by: str) -> EmergencyAllocationResponse: ...
    def get_summary(self) -> EmergencySummary: ...


# ---------------------------------------------------------------------------
# In-memory
# ---------------------------------------------------------------------------


class InMemoryEmergencyRepository:
    """Deterministic in-memory emergency repository for demo mode."""

    def __init__(self) -> None:
        pass

    def _requests(self) -> dict[str, EmergencyRequestData]:
        return state.emergency_requests

    def create_request(self, consumer_id: str, consumer_name: str, data: Any) -> EmergencyRequestResponse:
        req = EmergencyRequestData(
            consumer_id=consumer_id,
            consumer_name=consumer_name,
            title=data.title,
            category=data.category.value,
            description=data.description,
            required_energy_kwh=data.required_energy_kwh,
            priority=data.priority.value,
            latitude=data.latitude,
            longitude=data.longitude,
            address=data.address,
            phone=data.phone,
            image_url=data.image_url,
            admin_notes=getattr(data, 'notes', None),
        )
        self._requests()[req.id] = req
        return _request_to_response(req)

    def get_my_requests(self, consumer_id: str) -> list[EmergencyRequestResponse]:
        return [
            _request_to_response(req)
            for req in self._requests().values()
            if req.consumer_id == consumer_id
        ]

    def get_all_requests(self) -> list[EmergencyRequestResponse]:
        return [_request_to_response(req) for req in self._requests().values()]

    def get_request(self, request_id: str) -> EmergencyRequestResponse | None:
        req = self._requests().get(request_id)
        return _request_to_response(req) if req else None

    def update_request(self, request_id: str, update: Any, admin_id: str) -> EmergencyRequestResponse:
        req = self._requests().get(request_id)
        if not req:
            raise ApiError(404, ErrorCode.RESOURCE_NOT_FOUND, f"Emergency request {request_id} not found")

        updates = {}
        if update.status:
            updates["status"] = update.status.value
            updates["assigned_admin"] = admin_id
            if update.status.value == "Approved":
                updates["approved_at"] = now_utc()
            if update.status.value == "Completed":
                updates["completed_at"] = now_utc()
        if update.admin_notes is not None:
            updates["admin_notes"] = update.admin_notes
        if update.allocated_energy_kwh is not None:
            updates["allocated_energy_kwh"] = update.allocated_energy_kwh

        updates["updated_at"] = now_utc()
        updated = req.model_copy(update=updates)
        self._requests()[request_id] = updated
        return _request_to_response(updated)

    def create_allocation(self, data: Any, allocated_by: str) -> EmergencyAllocationResponse:
        source_val = data.source.value
        alloc = EmergencyAllocationData(
            request_id=data.request_id,
            source=source_val,
            allocated_energy=data.allocated_energy,
            remarks=data.remarks,
            allocated_by=allocated_by,
        )
        req = self._requests().get(data.request_id)
        if req:
            req.allocations.append(alloc)
            current_allocated = req.allocated_energy_kwh
            req.allocated_energy_kwh = current_allocated + data.allocated_energy
            self._requests()[data.request_id] = req
        return EmergencyAllocationResponse(
            id=alloc.id,
            request_id=alloc.request_id,
            source=alloc.source,
            allocated_energy=alloc.allocated_energy,
            remarks=alloc.remarks,
            allocated_by=alloc.allocated_by,
            allocated_at=alloc.allocated_at,
        )

    def get_summary(self) -> EmergencySummary:
        total = 0
        pending = 0
        approved = 0
        rejected = 0
        completed = 0
        critical = 0
        for req in self._requests().values():
            total += 1
            if req.status == "Pending":
                pending += 1
            elif req.status == "Approved":
                approved += 1
            elif req.status == "Rejected":
                rejected += 1
            elif req.status == "Completed":
                completed += 1
            if req.priority == "Critical":
                critical += 1
        return EmergencySummary(
            total=total,
            pending=pending,
            approved=approved,
            rejected=rejected,
            completed=completed,
            critical=critical,
        )


# ---------------------------------------------------------------------------
# Supabase-backed
# ---------------------------------------------------------------------------


class SupabaseEmergencyRepository:
    """Emergency repository backed by Supabase."""

    def __init__(self, settings: Settings | None = None) -> None:
        current = settings or get_settings()
        self._client = get_supabase_admin_client(current)
        self._in_memory = InMemoryEmergencyRepository()

    def _require_client(self) -> None:
        if self._client is None:
            raise RuntimeError("Supabase is not configured for live mode.")

    def create_request(self, consumer_id: str, consumer_name: str, data: Any) -> EmergencyRequestResponse:
        self._require_client()
        try:
            payload = {
                "consumer_id": consumer_id,
                "title": data.title,
                "category": data.category.value,
                "description": data.description,
                "required_energy_kwh": data.required_energy_kwh,
                "priority": data.priority.value,
                "latitude": data.latitude,
                "longitude": data.longitude,
                "address": data.address,
                "phone": data.phone,
                "image_url": data.image_url,
                "admin_notes": getattr(data, 'notes', None),
            }
            result = self._client.table("emergency_requests").insert(payload).execute()
            if not result.data:
                return self._in_memory.create_request(consumer_id, consumer_name, data)
            row = result.data[0]
            return _row_to_response(row, consumer_name)
        except Exception as exc:
            logger.warning("Supabase emergency create failed, using in-memory: %s", exc)
            return self._in_memory.create_request(consumer_id, consumer_name, data)

    def get_my_requests(self, consumer_id: str) -> list[EmergencyRequestResponse]:
        self._require_client()
        try:
            result = self._client.table("emergency_requests")\
                .select("*")\
                .eq("consumer_id", consumer_id)\
                .order("created_at", desc=True)\
                .execute()
            rows = result.data or []
            responses = []
            for row in rows:
                consumer_name = self._get_user_name(row.get("consumer_id", ""))
                responses.append(_row_to_response(row, consumer_name))
            return responses
        except Exception as exc:
            logger.warning("Supabase emergency get_my_requests failed, using in-memory: %s", exc)
            return self._in_memory.get_my_requests(consumer_id)

    def get_all_requests(self) -> list[EmergencyRequestResponse]:
        self._require_client()
        try:
            result = self._client.table("emergency_requests")\
                .select("*")\
                .order("created_at", desc=True)\
                .execute()
            rows = result.data or []
            responses = []
            for row in rows:
                consumer_name = self._get_user_name(row.get("consumer_id", ""))
                responses.append(_row_to_response(row, consumer_name))
            return responses
        except Exception as exc:
            logger.warning("Supabase emergency get_all_requests failed, using in-memory: %s", exc)
            return self._in_memory.get_all_requests()

    def get_request(self, request_id: str) -> EmergencyRequestResponse | None:
        self._require_client()
        try:
            result = self._client.table("emergency_requests")\
                .select("*")\
                .eq("id", request_id)\
                .execute()
            if not result.data:
                return self._in_memory.get_request(request_id)
            row = result.data[0]
            consumer_name = self._get_user_name(row.get("consumer_id", ""))
            return _row_to_response(row, consumer_name)
        except Exception as exc:
            logger.warning("Supabase emergency get_request failed, using in-memory: %s", exc)
            return self._in_memory.get_request(request_id)

    def update_request(self, request_id: str, update: Any, admin_id: str) -> EmergencyRequestResponse:
        self._require_client()
        try:
            payload: dict[str, object] = {"updated_at": now_utc().isoformat()}
            if update.status:
                payload["status"] = update.status.value
                payload["assigned_admin"] = admin_id
                if update.status.value == "Approved":
                    payload["approved_at"] = now_utc().isoformat()
                if update.status.value == "Completed":
                    payload["completed_at"] = now_utc().isoformat()
            if update.admin_notes is not None:
                payload["admin_notes"] = update.admin_notes
            if update.allocated_energy_kwh is not None:
                payload["allocated_energy_kwh"] = update.allocated_energy_kwh

            result = self._client.table("emergency_requests")\
                .update(payload)\
                .eq("id", request_id)\
                .execute()
            if not result.data:
                return self._in_memory.update_request(request_id, update, admin_id)
            row = result.data[0]
            consumer_name = self._get_user_name(row.get("consumer_id", ""))
            return _row_to_response(row, consumer_name)
        except Exception as exc:
            logger.warning("Supabase emergency update_request failed, using in-memory: %s", exc)
            return self._in_memory.update_request(request_id, update, admin_id)

    def create_allocation(self, data: Any, allocated_by: str) -> EmergencyAllocationResponse:
        self._require_client()
        try:
            source_val = data.source.value
            payload = {
                "request_id": data.request_id,
                "source": source_val,
                "allocated_energy": data.allocated_energy,
                "remarks": data.remarks,
                "allocated_by": allocated_by,
            }
            result = self._client.table("emergency_allocations").insert(payload).execute()
            if result.data:
                row = result.data[0]
                return EmergencyAllocationResponse(
                    id=row.get("id", ""),
                    request_id=row.get("request_id", ""),
                    source=row.get("source", ""),
                    allocated_energy=float(row.get("allocated_energy", 0)),
                    remarks=row.get("remarks"),
                    allocated_by=row.get("allocated_by", ""),
                    allocated_at=_parse_dt(row.get("allocated_at")),
                )
        except Exception as exc:
            logger.warning("Supabase emergency create_allocation failed, using in-memory: %s", exc)
        return self._in_memory.create_allocation(data, allocated_by)

    def get_summary(self) -> EmergencySummary:
        self._require_client()
        try:
            all_reqs = self._client.table("emergency_requests").select("status, priority").execute()
            rows = all_reqs.data or []
            total = len(rows)
            pending = sum(1 for r in rows if r.get("status") == "Pending")
            approved = sum(1 for r in rows if r.get("status") == "Approved")
            rejected = sum(1 for r in rows if r.get("status") == "Rejected")
            completed = sum(1 for r in rows if r.get("status") == "Completed")
            critical = sum(1 for r in rows if r.get("priority") == "Critical")
            return EmergencySummary(
                total=total, pending=pending, approved=approved,
                rejected=rejected, completed=completed, critical=critical,
            )
        except Exception as exc:
            logger.warning("Supabase emergency get_summary failed, using in-memory: %s", exc)
            return self._in_memory.get_summary()

    def _get_user_name(self, user_id: str) -> str:
        try:
            result = self._client.table("profiles").select("full_name").eq("id", user_id).execute()
            if result.data:
                return result.data[0].get("full_name", "") or ""
        except Exception:
            pass
        return ""


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def _request_to_response(req: EmergencyRequestData) -> EmergencyRequestResponse:
    return EmergencyRequestResponse(
        id=req.id,
        consumer_id=req.consumer_id,
        consumer_name=req.consumer_name,
        title=req.title,
        category=req.category,
        description=req.description,
        required_energy_kwh=req.required_energy_kwh,
        allocated_energy_kwh=req.allocated_energy_kwh,
        priority=req.priority,
        status=req.status,
        latitude=req.latitude,
        longitude=req.longitude,
        address=req.address,
        phone=req.phone,
        image_url=req.image_url,
        admin_notes=req.admin_notes,
        assigned_admin=req.assigned_admin,
        created_at=req.created_at,
        updated_at=req.updated_at,
        approved_at=req.approved_at,
        completed_at=req.completed_at,
    )


def _row_to_response(row: dict, consumer_name: str = "") -> EmergencyRequestResponse:
    return EmergencyRequestResponse(
        id=row.get("id", ""),
        consumer_id=row.get("consumer_id", ""),
        consumer_name=consumer_name,
        title=row.get("title", ""),
        category=row.get("category", ""),
        description=row.get("description", ""),
        required_energy_kwh=float(row.get("required_energy_kwh", 0)),
        allocated_energy_kwh=float(row.get("allocated_energy_kwh", 0)),
        priority=row.get("priority", "Medium"),
        status=row.get("status", "Pending"),
        latitude=_float_or_none(row.get("latitude")),
        longitude=_float_or_none(row.get("longitude")),
        address=row.get("address"),
        phone=row.get("phone"),
        image_url=row.get("image_url"),
        admin_notes=row.get("admin_notes"),
        assigned_admin=row.get("assigned_admin"),
        created_at=_parse_dt(row.get("created_at")),
        updated_at=_parse_dt(row.get("updated_at")),
        approved_at=_parse_dt(row.get("approved_at")),
        completed_at=_parse_dt(row.get("completed_at")),
    )


def _parse_dt(val: object) -> Any:
    from datetime import datetime
    if isinstance(val, str):
        try:
            return datetime.fromisoformat(val.replace("Z", "+00:00"))
        except ValueError:
            return datetime.utcnow()
    if isinstance(val, datetime):
        return val
    return datetime.utcnow()


def _float_or_none(val: object) -> float | None:
    if val is None:
        return None
    try:
        return float(val)
    except (ValueError, TypeError):
        return None


# ---------------------------------------------------------------------------
# Factory
# ---------------------------------------------------------------------------


def get_emergency_repository(settings: Settings | None = None) -> EmergencyRepository:
    current = settings or get_settings()
    if current.supabase_url and current.supabase_service_role_key:
        return SupabaseEmergencyRepository(current)
    return InMemoryEmergencyRepository()

File: frontend/lib/features/emergency/data/emergency_repository.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../../core/config/app_config.dart';
import '../../../core/network/api_client.dart';
import 'emergency_models.dart';

/// Abstract repository for emergency assistance.
abstract class EmergencyRepository {
  Future<List<EmergencyRequest>> getMyRequests();
  Future<EmergencyRequest> createRequest(Map<String, dynamic> data);
  Future<EmergencyRequest> getRequest(String requestId);

  // Admin
  Future<List<EmergencyRequest>> getAllRequests();
  Future<EmergencyRequest> updateRequest(
    String requestId,
    Map<String, dynamic> data,
  );
  Future<EmergencyAllocation> createAllocation(Map<String, dynamic> data);
  Future<EmergencySummary> getSummary();
}

/// Provider that selects mock or live repository based on USE_MOCK_BACKEND.
final emergencyRepositoryProvider = Provider<EmergencyRepository>((ref) {
  if (ref.watch(appConfigProvider).isLiveMode) {
    return EmergencyApiRepository(ref.watch(apiClientProvider));
  }
  return EmergencyMockRepository();
});

/// Live API implementation.
class EmergencyApiRepository implements EmergencyRepository {
  EmergencyApiRepository(this._client);

  final ApiClient _client;

  @override
  Future<List<EmergencyRequest>> getMyRequests() async {
    final data = await _client.get('/emergency/mine');
    return (data as List)
        .map((e) => EmergencyRequest.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  @override
  Future<EmergencyRequest> createRequest(Map<String, dynamic> data) async {
    final result = await _client.post('/emergency', body: data);
    return EmergencyRequest.fromJson(result as Map<String, dynamic>);
  }

  @override
  Future<EmergencyRequest> getRequest(String requestId) async {
    final result = await _client.get('/emergency/$requestId');
    return EmergencyRequest.fromJson(result as Map<String, dynamic>);
  }

  @override
  Future<List<EmergencyRequest>> getAllRequests() async {
    final data = await _client.get('/admin/emergency/requests');
    return (data as List)
        .map((e) => EmergencyRequest.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  @override
  Future<EmergencyRequest> updateRequest(
    String requestId,
    Map<String, dynamic> data,
  ) async {
    final result = await _client.patch(
      '/admin/emergency/requests/$requestId',
      body: data,
    );
    return EmergencyRequest.fromJson(result as Map<String, dynamic>);
  }

  @override
  Future<EmergencyAllocation> createAllocation(
    Map<String, dynamic> data,
  ) async {
    final result = await _client.post(
      '/admin/emergency/allocations',
      body: data,
    );
    return EmergencyAllocation.fromJson(result as Map<String, dynamic>);
  }

  @override
  Future<EmergencySummary> getSummary() async {
    final result = await _client.get('/admin/emergency/summary');
    return EmergencySummary.fromJson(result as Map<String, dynamic>);
  }
}

/// Deterministic mock implementation.
class EmergencyMockRepository implements EmergencyRepository {
  EmergencyMockRepository();

  final List<EmergencyRequest> _mockRequests = [];
  int _idCounter = 0;

  @override
  Future<List<EmergencyRequest>> getMyRequests() async {
    await Future.delayed(const Duration(milliseconds: 200));
    return _mockRequests.where((r) => r.consumerId == 'current-user').toList();
  }

  @override
  Future<EmergencyRequest> createRequest(Map<String, dynamic> data) async {
    await Future.delayed(const Duration(milliseconds: 300));
    _idCounter++;
    final now = DateTime.now();
    final request = EmergencyRequest(
      id: 'EMR-MOCK-${_idCounter.toString().padLeft(4, '0')}',
      consumerId: 'current-user',
      consumerName: 'Test Consumer',
      title: data['title'] as String? ?? '',
      category: data['category'] as String? ?? 'Other',
      description: data['description'] as String? ?? '',
      requiredEnergyKwh: (data['required_energy_kwh'] as num?)?.toDouble() ?? 0,
      priority: data['priority'] as String? ?? 'Medium',
      status: 'Pending',
      latitude: data['latitude'] as double?,
      longitude: data['longitude'] as double?,
      address: data['address'] as String?,
      phone: data['phone'] as String?,
      imageUrl: data['image_url'] as String?,
      createdAt: now,
      updatedAt: now,
    );
    _mockRequests.insert(0, request);
    return request;
  }

  @override
  Future<EmergencyRequest> getRequest(String requestId) async {
    await Future.delayed(const Duration(milliseconds: 100));
    return _mockRequests.firstWhere(
      (r) => r.id == requestId,
      orElse: () => _createDemoRequest(),
    );
  }

  @override
  Future<List<EmergencyRequest>> getAllRequests() async {
    await Future.delayed(const Duration(milliseconds: 200));
    if (_mockRequests.isEmpty) {
      _seedRequests();
    }
    return List.from(_mockRequests);
  }

  @override
  Future<EmergencyRequest> updateRequest(
    String requestId,
    Map<String, dynamic> data,
  ) async {
    await Future.delayed(const Duration(milliseconds: 200));
    final index = _mockRequests.indexWhere((r) => r.id == requestId);
    if (index == -1) throw Exception('Request not found');

    final existing = _mockRequests[index];
    final updatedJson = {
      'id': existing.id,
      'consumer_id': existing.consumerId,
      'consumer_name': existing.consumerName,
      'title': existing.title,
      'category': existing.category,
      'description': existing.description,
      'required_energy_kwh': existing.requiredEnergyKwh,
      'allocated_energy_kwh':
          data['allocated_energy_kwh'] ?? existing.allocatedEnergyKwh,
      'priority': existing.priority,
      'status': data['status'] ?? existing.status,
      'latitude': existing.latitude,
      'longitude': existing.longitude,
      'address': existing.address,
      'phone': existing.phone,
      'admin_notes': data['admin_notes'] ?? existing.adminNotes,
      'created_at': existing.createdAt.toIso8601String(),
      'updated_at': DateTime.now().toIso8601String(),
    };

    // Update status-based timestamps
    if (data['status'] == 'Approved') {
      updatedJson['approved_at'] = DateTime.now().toIso8601String();
    }
    if (data['status'] == 'Completed') {
      updatedJson['completed_at'] = DateTime.now().toIso8601String();
    }

    final updated = EmergencyRequest.fromJson(updatedJson);
    _mockRequests[index] = updated;
    return updated;
  }

  @override
  Future<EmergencyAllocation> createAllocation(
    Map<String, dynamic> data,
  ) async {
    await Future.delayed(const Duration(milliseconds: 200));
    return EmergencyAllocation(
      id: 'EMA-MOCK-${(_idCounter++).toString().padLeft(4, '0')}',
      requestId: data['request_id'] as String? ?? '',
      source: data['source'] as String? ?? 'Government Reserve',
      allocatedEnergy: (data['allocated_energy'] as num?)?.toDouble() ?? 0,
      remarks: data['remarks'] as String?,
      allocatedBy: 'admin-current',
      allocatedAt: DateTime.now(),
    );
  }

  @override
  Future<EmergencySummary> getSummary() async {
    await Future.delayed(const Duration(milliseconds: 100));
    if (_mockRequests.isEmpty) {
      _seedRequests();
    }
    int pending = 0, approved = 0, rejected = 0, completed = 0, critical = 0;
    for (final r in _mockRequests) {
      switch (r.status) {
        case 'Pending':
          pending++;
          break;
        case 'Approved':
          approved++;
          break;
        case 'Rejected':
          rejected++;
          break;
        case 'Completed':
          completed++;
          break;
      }
      if (r.isCritical) critical++;
    }
    return EmergencySummary(
      total: _mockRequests.length,
      pending: pending,
      approved: approved,
      rejected: rejected,
      completed: completed,
      critical: critical,
    );
  }

  EmergencyRequest _createDemoRequest() {
    return EmergencyRequest(
      id: 'EMR-MOCK-0001',
      consumerId: 'current-user',
      consumerName: 'Ananya Nair',
      title: 'Medical Emergency - Life Support Equipment',
      category: 'Medical',
      description:
          'Need urgent power for life support equipment at home due to grid failure.',
      requiredEnergyKwh: 15.0,
      priority: 'Critical',
      status: 'Approved',
      address: '123 Main St, Kochi',
      phone: '+91-9876543210',
      createdAt: DateTime.now().subtract(const Duration(hours: 3)),
      updatedAt: DateTime.now().subtract(const Duration(minutes: 30)),
      approvedAt: DateTime.now().subtract(const Duration(hours: 2)),
    );
  }

  void _seedRequests() {
    final now = DateTime.now();
    _mockRequests.addAll([
      EmergencyRequest(
        id: 'EMR-MOCK-0001',
        consumerId: 'consumer-001',
        consumerName: 'Ananya Nair',
        title: 'Medical Emergency - Life Support Equipment',
        category: 'Medical',
        description:
            'Need urgent power for life support equipment at home due to grid failure.',
        requiredEnergyKwh: 15.0,
        allocatedEnergyKwh: 15.0,
        priority: 'Critical',
        status: 'Completed',
        address: '123 Main St, Kochi',
        phone: '+91-9876543210',
        createdAt: now.subtract(const Duration(hours: 24)),
        updatedAt: now.subtract(const Duration(hours: 1)),
        approvedAt: now.subtract(const Duration(hours: 23)),
        completedAt: now.subtract(const Duration(hours: 1)),
      ),
      EmergencyRequest(
        id: 'EMR-MOCK-0002',
        consumerId: 'consumer-002',
        consumerName: 'Biju Mathew',
        title: 'Flood Relief - Community Shelter',
        category: 'Flood',
        description:
            'Power needed for community shelter housing 50 people after flooding.',
        requiredEnergyKwh: 50.0,
        allocatedEnergyKwh: 40.0,
        priority: 'Critical',
        status: 'Approved',
        address: 'Flood Relief Camp, Kozhikode',
        phone: '+91-9876543211',
        latitude: 11.2588,
        longitude: 75.7804,
        createdAt: now.subtract(const Duration(hours: 6)),
        updatedAt: now.subtract(const Duration(hours: 2)),
        approvedAt: now.subtract(const Duration(hours: 4)),
      ),
      EmergencyRequest(
        id: 'EMR-MOCK-0003',
        consumerId: 'consumer-003',
        consumerName: 'Priya Sharma',
        title: 'Hospital Backup Power',
        category: 'Hospital',
        description:
            'Backup power required for critical medical equipment during maintenance.',
        requiredEnergyKwh: 100.0,
        priority: 'High',
        status: 'Pending',
        address: 'City Hospital, Kochi',
        phone: '+91-9876543212',
        createdAt: now.subtract(const Duration(hours: 1)),
        updatedAt: now.subtract(const Duration(hours: 1)),
      ),
      EmergencyRequest(
        id: 'EMR-MOCK-0004',
        consumerId: 'consumer-004',
        consumerName: 'Ravi Krishnan',
        title: 'Fire Rescue Support',
        category: 'Fire',
        description:
            'Emergency power for fire rescue operations at commercial building.',
        requiredEnergyKwh: 25.0,
        priority: 'Critical',
        status: 'In Progress',
        address: 'Market Complex, Thrissur',
        phone: '+91-9876543213',
        createdAt: now.subtract(const Duration(minutes: 45)),
        updatedAt: now.subtract(const Duration(minutes: 15)),
        approvedAt: now.subtract(const Duration(minutes: 30)),
      ),
      EmergencyRequest(
        id: 'EMR-MOCK-0005',
        consumerId: 'consumer-005',
        consumerName: 'Deepak Menon',
        title: 'Generator Failure - Medical Equipment',
        category: 'Medical',
        description:
            'Primary generator failed. Need immediate power backup for medical devices.',
        requiredEnergyKwh: 10.0,
        priority: 'Critical',
        status: 'Pending',
        address: 'Green Valley, Thrissur',
        phone: '+91-9876543214',
        createdAt: now.subtract(const Duration(minutes: 10)),
        updatedAt: now.subtract(const Duration(minutes: 10)),
      ),
    ]);
  }
}
